include/swift/AST/Decl.h

```diff
@@ -5454,6 +5421,123 @@ class FuncDecl final : public AbstractFunctionDecl,
+                              AccessorKind accessorKind,
+                              AddressorKind addressorKind,
+                              AbstractStorageDecl *storage,
+                              SourceLoc staticLoc,
+                              StaticSpellingKind staticSpelling,
+                              bool throws, SourceLoc throwsLoc,
+                              GenericParamList *genericParams,
+                              unsigned numParameterLists,
+                              DeclContext *parent);
+
+  static AccessorDecl *create(ASTContext &ctx, SourceLoc declLoc,
+                              SourceLoc accessorKeywordLoc,
+                              AccessorKind accessorKind,
+                              AddressorKind addressorKind,
+                              AbstractStorageDecl *storage,
+                              SourceLoc staticLoc,
+                              StaticSpellingKind staticSpelling,
+                              bool throws, SourceLoc throwsLoc,
+                              GenericParamList *genericParams,
+                              ArrayRef<ParameterList *> parameterLists,
+                              TypeLoc fnRetType, DeclContext *parent,
+                              ClangNode clangNode = ClangNode());
+
+  SourceLoc getAccessorKeywordLoc() const { return AccessorKeywordLoc; }
+
+  AbstractStorageDecl *getStorage() const {
+    return Storage;
+  }
+
+  AccessorKind getAccessorKind() const {
+    return AccessorKind(Bits.AccessorDecl.AccessorKind);
+  }
+
+  AddressorKind getAddressorKind() const {
+    return AddressorKind(Bits.AccessorDecl.AddressorKind);
+  }
+
+  bool isGetter() const { return getAccessorKind() == AccessorKind::IsGetter; }
+  bool isSetter() const { return getAccessorKind() == AccessorKind::IsSetter; }
+  bool isMaterializeForSet() const {
+    return getAccessorKind() == AccessorKind::IsMaterializeForSet;
+  }
+  bool isAnyAddressor() const {
+    auto kind = getAccessorKind();
+    return kind == AccessorKind::IsAddressor
+        || kind == AccessorKind::IsMutableAddressor;
+  }
+
+  /// isGetterOrSetter - Determine whether this is a getter or a setter vs.
+  /// a normal function.
+  bool isGetterOrSetter() const { return isGetter() || isSetter(); }
+  bool isObservingAccessor() const {
+    return getAccessorKind() == AccessorKind::IsDidSet ||
+           getAccessorKind() == AccessorKind::IsWillSet;
+  }
+
+  static bool classof(const Decl *D) {
+    return D->getKind() == DeclKind::Accessor;
+  }
+  static bool classof(const AbstractFunctionDecl *D) {
+    return classof(static_cast<const Decl*>(D));
+  }
+  static bool classof(const DeclContext *DC) {
+    if (auto fn = dyn_cast<AbstractFunctionDecl>(DC))
+      return classof(fn);
+    return false;
+  }
+};
+
+inline ParameterList **FuncDecl::getParameterListBuffer() {
+  if (!isa<AccessorDecl>(this))
+    return reinterpret_cast<ParameterList**>(this+1);
```

Comment:
Can you assert that `getKind()` is actually `Func`, and that we haven't added another subclass of FuncDecl?

Reply:
Sure thing.